controllers/idler/idler_controller.go

```diff
@@ -381,71 +384,73 @@ func (r *Reconciler) getUserEmailsFromMURs(ctx context.Context, hostCluster *clu
 	return emails, nil
 }
 
-// scaleControllerToZero checks if the object has an owner controller (Deployment, ReplicaSet, etc)
-// and scales the owner down to zero and returns "true".
-// Otherwise, returns "false".
-// It also returns the parent controller type and name or empty strings if there is no parent controller.
-func (r *Reconciler) scaleControllerToZero(ctx context.Context, meta metav1.ObjectMeta) (string, string, bool, error) {
-	log.FromContext(ctx).Info("Scaling controller to zero", "name", meta.Name)
-	owners := meta.GetOwnerReferences()
-	for _, owner := range owners {
-		if owner.Controller != nil && *owner.Controller {
-			switch owner.Kind {
-			case "Deployment":
-				return r.scaleDeploymentToZero(ctx, meta.Namespace, owner)
-			case "ReplicaSet":
-				return r.scaleReplicaSetToZero(ctx, meta.Namespace, owner)
-			case "DaemonSet":
-				return r.deleteDaemonSet(ctx, meta.Namespace, owner) // Nothing to scale down. Delete instead.
-			case "StatefulSet":
-				return r.scaleStatefulSetToZero(ctx, meta.Namespace, owner)
-			case "DeploymentConfig":
-				return r.scaleDeploymentConfigToZero(ctx, meta.Namespace, owner)
-			case "ReplicationController":
-				return r.scaleReplicationControllerToZero(ctx, meta.Namespace, owner)
-			case "Job":
-				return r.deleteJob(ctx, meta.Namespace, owner) // Nothing to scale down. Delete instead.
-			case "VirtualMachineInstance":
-				return r.stopVirtualMachine(ctx, meta.Namespace, owner) // Nothing to scale down. Stop instead.
-			}
-		}
-	}
-	return "", "", false, nil
+// scaleControllerToZero fetches the whole tree of the controller owners from the provided object (Deployment, ReplicaSet, etc).
+// If any known controller owner is found, then it's scaled down (or deleted) and its kind and name is returned.
+// Otherwise, returns empty strings.
+func (r *Reconciler) scaleControllerToZero(ctx context.Context, meta metav1.Object, ownerFetcher *ownerFetcher) (kind string, name string, err error) {
+	log.FromContext(ctx).Info("Scaling controller to zero", "name", meta.GetName())
+	var owners []*objectWithGVR
+	owners, err = ownerFetcher.getOwners(ctx, meta)
+	if err != nil {
+		return
+	}
+	for _, ownerWithGVR := range owners {
+		owner := ownerWithGVR.object
+		kind = owner.GetObjectKind().GroupVersionKind().Kind
```

Comment:
Does the dynamic client fill in the `TypeMeta` of an object? Because generally, one cannot assume it is filled in in a `client.Object` instance. E.g. the dance in https://github.com/codeready-toolchain/toolchain-common/blob/master/pkg/client/ssa_client.go#L223

Reply:
it's a resource retrieved from a cluster (the type is unstructured). I expect that any resource that is retrieved from the cluster/etcd with any type of client should contain the TypeMeta.
In your case (the line that you referenced) the code can read objects that were created in the code (using the specific type of the resource) without setting the TypeMeta explicitly

Reply:
> apart from that, the code will work with resources whose API is not in Scheme (eg. AAP or VMs)

That's a very good point.

Should we think about modifying our apply clients to work explicitly with `Unstructured` and dynamic client, too, so that we're not bound to a small subset of resources that we "know about"?

Reply:
I think that the SSA code is written in a way it can support both:
* `Unstructured` (or some generic `runtime.Object`) when it's retrieved from a template/cluster - in this case it contains the `TypeMeta`. In this case, the API doesn't have to be in the Schema
* typed objects that don't have the `TypeMeta` but it's API is in the Schema, so the logic can automatically fill it in

this is exactly what we want